Source/Renderer/FramebufferManager.js

```diff
@@ -0,0 +1,308 @@
+ *
+ * @exception {DeveloperError} Cannot have both a depth and depth-stencil attachment.
+ *
+ * @private
+ * @constructor
+ */
+function FramebufferManager(options) {
+  options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+  this._colorAttachmentsLength = defaultValue(
+    options.colorAttachmentsLength,
+    1
+  );
+
+  this._color = defaultValue(options.color, true);
+  this._depth = defaultValue(options.depth, false);
+  this._depthStencil = defaultValue(options.depthStencil, false);
+  //>>includeStart('debug', pragmas.debug);
+  if (this._depth && this._depthStencil) {
+    throw new DeveloperError(
+      "Cannot have both a depth and depth-stencil attachment."
+    );
+  }
+  //>>includeEnd('debug');
+
+  this._createColorAttachments = defaultValue(
+    options.createColorAttachments,
+    true
+  );
+  this._createDepthAttachments = defaultValue(
+    options.createDepthAttachments,
+    true
+  );
+
+  this._useHdr = false;
+
+  this._framebuffer = undefined;
+  this._colorRenderbuffer = undefined;
+  this._colorTextures = [];
+  this._depthStencilRenderbuffer = undefined;
+  this._depthStencilTexture = undefined;
+  this._depthRenderbuffer = undefined;
+  this._depthTexture = undefined;
+
+  this._attachmentsSet = false;
+}
+
+Object.defineProperties(FramebufferManager.prototype, {
+  framebuffer: {
+    get: function () {
+      return this._framebuffer;
+    },
+  },
+  status: {
+    get: function () {
+      return this._framebuffer.status;
+    },
+  },
+});
+
+FramebufferManager.prototype.isDirty = function (width, height, hdr) {
+  hdr = defaultValue(hdr, false);
+
+  var texturesDirty = false;
+  var length = this._colorTextures.length;
+  var texture;
+  for (var i = 0; i < length; ++i) {
+    texture = this._colorTextures[i];
+    if (
+      !defined(texture) ||
+      texture.width !== width ||
+      texture.height !== height
+    ) {
+      texturesDirty = true;
+      break;
+    }
+  }
+  return length === 0 || texturesDirty || this._useHdr !== hdr;
+};
+
+FramebufferManager.prototype.update = function (
+  context,
+  width,
+  height,
+  depthTexture,
+  hdr
+) {
+  //>>includeStart('debug', pragmas.debug);
+  if (
+    (!defined(width) || !defined(height)) &&
+    ((this._color && this._createColorAttachments) ||
+      ((this._depth || this._depthStencil) && this._createDepthAttachments))
+  ) {
+    throw new DeveloperError(
+      "width and height must be provided if color or depth attachments are created."
+    );
+  }
+  if (!this._color && !this._depth && !this._depthStencil) {
+    throw new DeveloperError(
+      "must enable at least one type of framebuffer attachment."
+    );
+  }
+  //>>includeEnd('debug');
+  depthTexture = defaultValue(depthTexture, false);
+  hdr = defaultValue(hdr, false);
+
+  if (this._attachmentsSet || this.isDirty(width, height, hdr)) {
+    if (!this._attachmentsSet) {
```

Comment:
This check makes me unsure if `attachmentsSet` should be part of the class at all. If `createXAttachments` being set to false for any type of attachment means that the manager can expect that attachment to have been set before `update`, I think we can get rid of this member variable.

Reply:
Resolved by removing the `attachmentsSet` variable.